python/paddle/fluid/dygraph/dygraph_to_static/ast_utils.py

```diff
@@ -0,0 +1,323 @@
+    control flow in Paddle.
+    """
+    return True
+
+
+def get_name_ids(nodes, not_name_set=None, node_black_list=None):
+    """
+    Return all ast.Name.id of python variable in nodes.
+    """
+    if not isinstance(nodes, (list, tuple, set)):
+        raise ValueError(
+            "nodes must be one of list, tuple, set, but received %s" %
+            type(nodes))
+    if not_name_set is None:
+        not_name_set = set()
+
+    def update(old_dict, new_dict):
+        for k, v in new_dict.items():
+            old_dict[k].extend(v)
+
+    name_ids = defaultdict(list)
+    for node in nodes:
+        if node_black_list and node in node_black_list: continue
+        if isinstance(node, gast.AST):
+            # In two case, the ast.Name should be filtered.
+            # 1. Function name like `my_func` of my_func(x)
+            # 2. api prefix like `fluid` of `fluid.layers.mean`
+            if isinstance(node, gast.Return):
+                continue
+            elif isinstance(node, gast.Call) and isinstance(node.func,
+                                                            gast.Name):
+                not_name_set.add(node.func.id)
+            elif isinstance(node, gast.Attribute) and isinstance(node.value,
+                                                                 gast.Name):
+                not_name_set.add(node.value.id)
+            if isinstance(
+                    node, gast.Name
+            ) and node.id not in name_ids and node.id not in not_name_set:
+                if isinstance(node.ctx, (gast.Store, gast.Load, gast.Param)):
+                    name_ids[node.id].append(node.ctx)
+            else:
+                if isinstance(node, gast.Assign):
+                    node = copy.copy(node)
+                    node._fields = ('value', 'targets')
+                for field, value in gast.iter_fields(node):
+                    value = value if isinstance(value, list) else [value]
+                    update(name_ids,
+                           get_name_ids(value, not_name_set, node_black_list))
+    return name_ids
+
+
+def parse_args(var_ids_dict, return_ids=None, ctx=gast.Load):
+    """
+    Find out the ast.Name.id list of input by analyzing node's AST information.
+    """
+
+    name_ids = [
+        var_id for var_id, var_ctx in var_ids_dict.items()
+        if isinstance(var_ctx[0], ctx)
+    ]
+    if return_ids:
+        new_args = set(return_ids) - set(name_ids)
+        name_ids.extend(list(new_args))
+    name_ids.sort()
+    args = [
+        gast.Name(
+            id=name_id, ctx=gast.Load(), annotation=None, type_comment=None)
+        for name_id in name_ids
+    ]
+    arguments = gast.arguments(
+        args=args,
+        posonlyargs=[],
+        vararg=None,
+        kwonlyargs=[],
+        kw_defaults=None,
+        kwarg=None,
+        defaults=[])
+    return arguments
+
+
+def parse_return(parent_vars_dict, if_vars_dict, else_vars_dict):
```

Comment:
Is this used for parsing return for cond? If yes, parse_cond_return may be better name.

Reply:
Thank, rename it into `parse_cond_return`